nullifier-merkle-tree/src/lib.rs

```diff
@@ -0,0 +1,414 @@
+use thiserror::Error;
+
+#[derive(Error, Debug)]
+pub enum NullifierMerkleTreeError {
+    #[error("Low nullifier not found")]
+    LowNullifierNotFound,
+}
+
+#[derive(Clone, Default, Debug, PartialEq)]
+pub struct Node {
+    value: [u8; 32],
+    next_index: usize,
+}
+
+pub struct NullifierMerkleTree<const HEIGHT: usize, const ROOTS: usize> {
+    pub nodes: [Node; HEIGHT],
+    pub roots: [Node; ROOTS],
+    pub current_node_index: usize,
+    pub current_root_index: usize,
+}
+
+impl<const HEIGHT: usize, const ROOTS: usize> Default for NullifierMerkleTree<HEIGHT, ROOTS> {
+    fn default() -> Self {
+        Self {
+            nodes: std::array::from_fn(|_| Node {
+                value: [0u8; 32],
+                next_index: 0,
+            }),
+            roots: std::array::from_fn(|_| Node {
+                value: [0u8; 32],
+                next_index: 0,
+            }),
+            current_node_index: 0,
+            current_root_index: 0,
+        }
+    }
+}
+
+impl<const HEIGHT: usize, const ROOTS: usize> NullifierMerkleTree<HEIGHT, ROOTS> {
+    fn find_low_nullifier(
+        &mut self,
+        new_value: &[u8; 32],
+    ) -> Result<usize, NullifierMerkleTreeError> {
+        for (i, nullifier) in self.nodes.iter().enumerate() {
+            if self.nodes[nullifier.next_index].value > *new_value {
+                return Ok(i);
+            }
+        }
+        for (i, nullifier) in self.nodes.iter().enumerate() {
+            if self.nodes[nullifier.next_index].value == [0u8; 32] {
+                return Ok(i);
+            }
+        }
+        Err(NullifierMerkleTreeError::LowNullifierNotFound)
+    }
+
+    pub fn insert(&mut self, nullifiers: &[&[u8; 32]]) -> Result<(), NullifierMerkleTreeError> {
+        for nullifier in nullifiers {
+            let low_nullifier_index = self.find_low_nullifier(nullifier)?;
+
+            let new_value = nullifier.to_owned().to_owned();
+
+            let new_node = Node {
+                value: new_value,
+                next_index: self.nodes[low_nullifier_index].next_index,
+            };
+
+            // Insert new node.
+            self.current_node_index += 1;
+            self.nodes[self.current_node_index] = new_node;
+
+            // Update the lower nullifier - point to the new node.
+            self.nodes[low_nullifier_index].next_index = self.current_node_index;
+
+            // Yield lower nullifier as a root.
+            self.current_root_index = (self.current_root_index + 1) % ROOTS;
+            self.roots[self.current_root_index] = self.nodes[low_nullifier_index].clone();
+
+            // Yield new node as a root.
+            self.current_root_index = (self.current_root_index + 1) % ROOTS;
+            self.roots[self.current_root_index] = self.nodes[self.current_node_index].clone();
```

Comment:
This seems incomplete to me, since you are not computing any hashes. The root hash should be the hashes of the the two child nodes which are the hashes of the two child nodes as in a regular binary Merkle tree if I understand correctly.
The only difference to a regular bin Merkle tree is what the leaves repesent and how to update the leaves. Once the leaves are updated the updated leaves need to be inserted into the binary Merkle tree.